Add possibility to split oversized udp batches

janLo opened the pull request:
# Description

If we use the BatchExportSpanProcessor combined with the
JaegerSpanExporter and use instrumentations that add a lot of metadata
to the spans like sqlalchemy, then we run occationally into the "Data
exceeds the max UDP packet size" warning causing dropped spans and
incomplete data.

The option to reduce the general batch-size to a very small number (in
my case >30) may cause a performance issue as the worker thread of the
batch exporter gets very busy. Instead this change allows the user to
ask the exporter to split oversized batches when they get detected and
send the splits separately instead of dropping them. Depending on the
usecase this is a better option than reducing the batch-size to a very
small value because every now and then they contain a couple of large
spans.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] A unittest was added
- [x] Its running successfully in production 

# Does This PR Require a Contrib Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

- [x] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

codeboten commented on exporter/opentelemetry-exporter-jaeger/src/opentelemetry/exporter/jaeger/__init__.py:
```diff
@@ -112,6 +113,7 @@ def __init__(
         collector_endpoint=None,
         username=None,
         password=None,
+        udp_split_oversized_batches=None,
```
This arg should be added to the docs above

codeboten commented on exporter/opentelemetry-exporter-jaeger/src/opentelemetry/exporter/jaeger/__init__.py:
```diff
@@ -140,13 +142,20 @@ def __init__(
         )
+        self.udp_split_oversized_batches = _parameter_setter(
+            param=udp_split_oversized_batches,
+            env_variable=Configuration().EXPORTER_JAEGER_AGENT_SPLIT_OVERSIZED_BATCHES,
```
Is this env variable defined in the spec? If not it would be great to open a PR. I didn't see it looking here: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/sdk-environment-variables.md#jaeger-exporter

codeboten commented on exporter/opentelemetry-exporter-jaeger/src/opentelemetry/exporter/jaeger/__init__.py:
```diff
@@ -386,6 +395,7 @@ class AgentClientUDP:
         port: The port of the Jaeger server.
         max_packet_size: Maximum size of UDP packet.
         client: Class for creating new client objects for agencies.
+        split_oversized_batches: Re-emit over
```
Is this comment incomplete? I'm not sure I understand it as-is.

codeboten reviewed (changes requested):
Thanks for the PR, it looks great. Just a couple of changes requested, please also add an entry to the changelog.

codeboten commented:
hey @janLo any chance you can take a look at the comments on this PR so we can get it merged?

janLo commented:
Yes, sorry, I'm just very busy at the moment. It's on my list and I plan to get it finished ASAP.

codeboten commented:
> Yes, sorry, I'm just very busy at the moment. It's on my list and I plan to get it finished ASAP.

No worries, thanks!

lzchen commented:
@janLo 
Any chance you are still working on this? The PR has been open for almost 2 months now.

janLo commented:
I hope to get to it next week.

janLo commented:
Sorry for the noise, I haven't had a working python env at hand today.

I fixed the mentioned issues and rebased everything on main. I also added a changelog entry and made a PR in the mentioned documentation repo: https://github.com/open-telemetry/opentelemetry-specification/pull/1475

bogdandrutu commented on exporter/opentelemetry-exporter-jaeger/src/opentelemetry/exporter/jaeger/__init__.py:
```diff
@@ -120,6 +121,7 @@ class JaegerSpanExporter(SpanExporter):
         insecure: True if collector has no encryption or authentication
         credentials: Credentials for server authentication.
         transport_format: Transport format for exporting spans to collector.
+        udp_split_oversized_batches: Re-emit oversized batches in smaller chunks.
```
Why only udp? Why not simply provide a size? But I am curios how this size is calculated.

janLo replied:
Because in the case of TCP your stack will just fragment your message.

seemk commented on exporter/opentelemetry-exporter-jaeger/src/opentelemetry/exporter/jaeger/send/thrift.py:
```diff
@@ -66,11 +70,23 @@ def emit(self, batch: jaeger.Batch):
+                    end = (packet + 1) * div
+                    self.emit(
+                        jaeger.Batch(
+                            process=batch.process, spans=batch.spans[start:end]
```
In some cases the last emit can produce a batch with empty spans? Can we avoid it? :slightly_smiling_face: 

For example, try with:
```
spans=4
buffer_length=131000
```

janLo replied:
Oh, thabk you! I've addressed this now.

janLo commented:
Can I somehow rerun teh failed check? I don't think it has anything to do with my changes.
